diff --git a/alvr/launcher/res/vrcompositor_launcher_wrapper.sh b/alvr/launcher/res/vrcompositor_launcher_wrapper.sh
index 85c369df..2566844c 100644
--- a/alvr/launcher/res/vrcompositor_launcher_wrapper.sh
+++ b/alvr/launcher/res/vrcompositor_launcher_wrapper.sh
@@ -1,6 +1,6 @@
 #!/usr/bin/env sh
 
-export VK_LAYER_PATH=$(cat $XDG_RUNTIME_DIR/alvr_dir.txt)/share/vulkan/explicit_layer.d
+export VK_LAYER_PATH="REPLACE_THIS_VK_LAYER_PATH"
 export VK_INSTANCE_LAYERS=VK_LAYER_ALVR_capture
 export DISABLE_VK_LAYER_VALVE_steam_fossilize_1=1
 
